### PR TITLE
rootless: keep CL_UNPRIVILEGED "locked" flags for bind-mounting resolv.conf

### PR DESCRIPTION
**Pull request by AkihiroSuda**
Port over moby/moby#35205

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

- - -

Without this, the mount fails EPERM on Ubuntu.

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc @cyphar

File: oci/layer/unpack.go [hunk 1]
```diff
@@ -309,7 +310,18 @@ func UnpackRuntimeJSON(ctx context.Context, engine cas.Engine, configFile io.Wri
 	}
 	if mapOptions.Rootless {
 		ToRootless(g.Spec())
-		g.AddBindMount("/etc/resolv.conf", "/etc/resolv.conf", []string{"bind", "ro"})
+		resolvConfSrc := "/etc/resolv.conf"
```
Comment (cyphar) [inline oci/layer/unpack.go]: This looks okay, but I would prefer if this was a `const`.

File: oci/layer/unpack.go [hunk 2]
```diff
@@ -309,7 +310,18 @@ func UnpackRuntimeJSON(ctx context.Context, engine cas.Engine, configFile io.Wri
...
+		// If we are using user namespaces, then we must make sure that we
+		// don't drop any of the CL_UNPRIVILEGED "locked" flags of the source
+		// "mount" when we bind-mount. The reason for this is that at the point
+		// when runc sets up the root filesystem, it is already inside a user
+		// namespace, and thus cannot change any flags that are locked.
+		unprivOpts, err := getUnprivilegedMountFlags(resolvConfSrc)
+		if err != nil {
+			return errors.Wrapf(err, "inspecting mount flags of %s", resolvConfSrc)
+		}
+		g.AddBindMount(resolvConfSrc, "/etc/resolv.conf",
```
Comment (cyphar) [inline oci/layer/unpack.go]: And you can use `resolvConfPath` for both arguments here.

**Comment by cyphar** [comment]
Some minor nits, otherwise looks fine.

**Comment by AkihiroSuda** [comment]
done

**Comment by cyphar** [comment]
LGTM. Will merge once tests pass.

**Pull request merged**